### DISCUSSION TITLE
Torch 2.0

### DISCUSSION THREAD

**Discussion by iDeNoh** [lshqqytiger/stable-diffusion-webui-directml#44] [General]
With the update to Pytorch 2.0 is there any way to update to it? I've tried manually updating using CLI and it seems to want to stay on 1.1.13, I tried removing my venv folder and adding " set TORCH_COMMAND= pip install torch torchvision pytorch-triton --extra-index-url https://download.pytorch.org/whl/cu118" to force an update and it seems that this breaks the install. Is it possible to update and try it out?

**Comment by lshqqytiger** [comment]
cu118 on url means CUDA 11.8.
Installation breaks because you don't have CUDA toolkit (if you don't have NVIDIA GPU).
In conclusion, DirectML can't use PyTorch 2.0 at this time.
DirectML for PyTorch is version 0.1.13.1 which means it is for PyTorch 1.13.1 doesn't support PyTorch 2.0.
We should wait for the next update on torch-directml that supports PyTorch 2.0.

**Reply by silvertuanzi** [reply]
torch-directml 0.2.0 which supports torch 2.0 was released 5 days ago. Plan to support?

**Reply by lshqqytiger** [reply]
I added torch 2.0 support while merging AUTOMATIC1111/master. [271edfa](https://github.com/lshqqytiger/stable-diffusion-webui-directml/commit/271edfa8d6b66e0d0b3f0b2964f35598df69d2a5)

**Reply by howDoILearnPython** [reply]
i tried to upgrade to torch 2.0 last week and got a bunch of error message, and eventually i revert back to cu116

Can we now upgrade to torch 2.0 and do i need to reinstall the webui-directML by git clone?

**Reply by lshqqytiger** [reply]
I think you don't need to reinstall it.
Just remove venv folder and pull latest commits.

**Reply by Milor123** [reply]
@lshqqytiger Bro, can u says me, how can I use directMl in other torch 2.0 projects? I am trying but i cant do it xD. Thank u very much, I am AMD user

**Reply by lshqqytiger** [reply]
It's very simple. Just move all the tensors to DirectML device. ([more examples/information](https://github.com/microsoft/DirectML/tree/master/PyTorch/1.13#setup))
But if you are not familiar with python and pytorch, it can be hard.

**Reply by Milor123** [reply]
Thank u very much, I am familiar with python. I try do it, my question is, 

1)on the code have .cpu(), it should convert in .to(dml)?  or only the 

![image](https://user-images.githubusercontent.com/14153649/236822843-e5637381-ff51-48d4-82c2-2325abc7eded.png)



2) Or only the **tensors with .to(device)**?

![image](https://user-images.githubusercontent.com/14153649/236823548-4d36d425-e03d-4a4d-8ca5-1025a36ccc58.png)

3) I've found .to(device) in other parts of the code, should i convert it?
![image](https://user-images.githubusercontent.com/14153649/236823779-fdf8f86e-06eb-4d8e-acc6-3f2e0844b1f6.png)

I've tried some things of this, and in some cases get error and another the GPU dont is used, only CPU. please help me bro !!

4) The project that i am editing use torch 2.0, should I **specify** in the next installations use **torch 1.13.1** ??

```
conda install numpy pandas tensorboard matplotlib tqdm pyyaml -y
pip install opencv-python
pip install wget
pip install torchvision
```

`conda install pytorch cpuonly -c pytorch`

Thank u very much for support me.

**Reply by lshqqytiger** [reply]
1. No.
2. Yes. But it will be better to find the source/origin/root of `device` and change it to DirectML device.
3. Yes.
4. No. Use torch 2.0. I think DirectML team did not edit their documents yet.

**Reply by Milor123** [reply]
Thank u very much! @lshqqytiger, now my GPU is active when try start, now have this problem, could you help me with some Idea to solve this. https://github.com/suno-ai/bark/issues/271

**Comment by Milor123** [comment]
@lshqqytiger Dear bro, Actually i am having the problem **`RuntimeError: Cannot set version_counter for inference tensor`** in the project that you help me. Try write to the developer but he saids that dont have experience in AMD.

I've changed all the **`.to(device)`** to **`.to(dml)`** according to [gpu-pytorch-windows Docs](https://learn.microsoft.com/en-us/windows/ai/directml/gpu-pytorch-windows) in the files **`generation.py`** in `bark` folder and `build\lib\bark\` respectly. When I try run the project. I seen that the GPU started correctly but then i get the next error.

I really appreciate all the help you've given me so far. I was hoping you could help me out again. I've been reading a lot and trying different things, but I can't find much information on this error I'm getting. I don't know what to do or if I'm doing something wrong.

> From here: https://github.com/suno-ai/bark/issues/271

```shell
python .\run.py
No GPU being used. Careful, inference might be very slow!
  0%|                                                                                                                                                | 0/100 [00:00<?, ?it/s]Traceback (most recent call last):
  File "C:\Users\NoeXVanitasXJunk\bark\run.py", line 13, in <module>
    audio_array = generate_audio(text_prompt)
  File "C:\Users\NoeXVanitasXJunk\bark\bark\api.py", line 107, in generate_audio
    semantic_tokens = text_to_semantic(
  File "C:\Users\NoeXVanitasXJunk\bark\bark\api.py", line 25, in text_to_semantic
    x_semantic = generate_text_semantic(
  File "C:\Users\NoeXVanitasXJunk\bark\bark\generation.py", line 460, in generate_text_semantic
    logits, kv_cache = model(
  File "C:\Users\NoeXVanitasXJunk\miniconda3\envs\tfdml_plugin\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Users\NoeXVanitasXJunk\bark\bark\model.py", line 208, in forward
    x, kv = block(x, past_kv=past_layer_kv, use_cache=use_cache)
  File "C:\Users\NoeXVanitasXJunk\miniconda3\envs\tfdml_plugin\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Users\NoeXVanitasXJunk\bark\bark\model.py", line 121, in forward
    attn_output, prev_kvs = self.attn(self.ln_1(x), past_kv=past_kv, use_cache=use_cache)
  File "C:\Users\NoeXVanitasXJunk\miniconda3\envs\tfdml_plugin\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Users\NoeXVanitasXJunk\bark\bark\model.py", line 50, in forward
    q, k ,v  = self.c_attn(x).split(self.n_embd, dim=2)
  File "C:\Users\NoeXVanitasXJunk\miniconda3\envs\tfdml_plugin\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Users\NoeXVanitasXJunk\miniconda3\envs\tfdml_plugin\lib\site-packages\torch\nn\modules\linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: Cannot set version_counter for inference tensor
  0%|                                                                                                                                                | 0/100 [00:00<?, ?it/s]
```

I am in **Python 3.9.16**



Could you try help me with this? 

Other thing: I've read about with **torch-mlir** is possible use AMD card, but not stay sure if on windows, I try it but i am not sure if need something more or some DirectMl Special
I try install **torch-mlir** and works in the project but it only use CPU not GPU, i am not sure how configure for using the GPU

**Reply by lshqqytiger** [reply]
I have met same error too. But I was unable to make it work.
torch-mlir is not as simple as DirectML (as I know)
but it works on Windows and has same or better performance compared to DirectML.

**Reply by Milor123** [reply]
@lshqqytiger  Thank u very much brother !!!! , I try add **`mode=False`** to **`inference_mode`**, then **`inference_mode(mode=False)`**, Now it seems to be working, it starts and loads but before carrying out its process, it tells me the following  `"UserWarning: The operator 'aten::tril.out' is not currently supported on the DML backend and will fall back to run on the CPU. This may have performance implications"`


```shell
No GPU being used. Careful, inference might be very slow!
  0%|                                                                                          | 0/100 [00:00<?, ?it/s]C:\Users\NoeXVanitasXJunk\bark\bark\model.py:80: UserWarning: The operator 'aten::tril.out' is not currently supported on the DML backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at D:\a\_work\1\s\pytorch-directml-plugin\torch_directml\csrc\dml\dml_cpu_fallback.cpp:17.)
  y = torch.nn.functional.scaled_dot_product_attention(q, k, v, dropout_p=self.dropout, is_causal=is_causal)
100%|████████████████████████████████████████████████████████████████████████████████| 100/100 [00:32<00:00,  3.04it/s]
  0%|                                                                                           | 0/31 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "C:\Users\NoeXVanitasXJunk\bark\run.py", line 13, in <module>
    audio_array = generate_audio(text_prompt)
  File "C:\Users\NoeXVanitasXJunk\bark\bark\api.py", line 113, in generate_audio
    out = semantic_to_waveform(
  File "C:\Users\NoeXVanitasXJunk\bark\bark\api.py", line 54, in semantic_to_waveform
    coarse_tokens = generate_coarse(
  File "C:\Users\NoeXVanitasXJunk\bark\bark\generation.py", line 633, in generate_coarse
    x_in = torch.hstack(
RuntimeError
```

**Comment by PennyFranklin** [comment]
i run into another problem. after updating to torch2.0, sdwebui uses almost all of the vram during the first picture generation, and then run at a very low speed when generate the second and more  picture

**Comment by FuzzyBootsDK** [comment]
I am getting this error.

Does anyone know how to fix it. 
i am on an AMD GPU
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/98462045/9f26a3c1-1879-4899-981f-5f092942b878)

**Reply by Milor123** [reply]
> I am getting this error.
> 
> Does anyone know how to fix it. i am on an AMD GPU ![image](https://user-images.githubusercontent.com/98462045/245434285-9f26a3c1-1879-4899-981f-5f092942b878.png)

Try use  [Python 3.10.6](https://www.python.org/downloads/release/python-3106/)

**Reply by FuzzyBootsDK** [reply]
I have removed all Python versions and installed 3.10.6 from the link, rebooted just to be sure.
Now i get this error, asking for 3.11 as i see it
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/98462045/813a25b0-f7ab-456a-bcd2-dcd3a38e48ff)

**Reply by Milor123** [reply]
> I have removed all Python versions and installed 3.10.6 from the link, rebooted just to be sure. Now i get this error, asking for 3.11 as i see it ![image](https://user-images.githubusercontent.com/98462045/245448044-813a25b0-f7ab-456a-bcd2-dcd3a38e48ff.png)

Ok very nice, now u need delete enviroment folder of the old python
delete the folder name **`venv`** in you main folder .\stable-diffusion-webui-directml\ **`venv`**
and run again the .bat file, the libraries will be installed again.
